keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -157,15 +162,24 @@ TODO: complete this list
 
 ### e2e tests for GA
 
-* Custom Resources should be readable and writable at all available versions (test for round trip-ability)
 * A Custom Resource watch should terminate if its CustomResourceDefinition is deleted or updated
 
 ### Scale Targets for GA
 
-* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
-* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
-* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
-* Coordinate with sig-scalability
+Scale targets here are both CRDs with and without a webhook converter.
+
+Conditions:
+
+- A trivial webhook converter with ~1ms processing latency.
+- Custom resource size is <10kb of data.
+- Kubernetes master components running on "standard" sized VM, e.g. GCE n1-standard-1.
+
+Targets:
+
+* Create latency of 200ms for CRDs  (within 50% of native Kubernetes types)
```

Comment:
We have the SLI/SLO for API call latencies of 1s:
https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md

Why we would like to try target something different here? We should be consistent (unless there is a very good reason which I'm not aware of).

Reply:
+1 for aligning with existing SLI/SLO. Two thoughts:
1. there are no object count limits accompanying the SLO targets (`(a) <= 1s if scope=resource (b) <= 5s if scope=namespace (c) <= 30s if scope=cluster`). We need *some* object count limit we can test at to ensure we are within SLO and to give users a sense of the scale that is tested/supported
2. would it make sense to express our SLO for custom resources which use webhook conversion in a way that either omits or assumes a fixed latency for the out-of-process conversion webhook? If we can measure the request handling time spent waiting for the conversion webhook, we can indicate our SLO, let custom resource webhook authors set their own SLO, and they can add the two together to get the total resource SLO.

Reply:
Re 1: I kind-of agree. There is this other doc that https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md that is talking about supported limits. I agree this is far from precise and not were we would like to be, but it's at least something.
In summary, I think that namespace currently translates to sth like <=5000 or so, and cluster translates to something fairly high (e.g. <=150k for pods).

Re 2: I completely agree. I'm leaning towards excluding (this would be consistent with e.g. SLI for pod startup time, where we exclude e.g. image pulling). And that would match the "two SLOs addition you had". But I'm open to other suggestions too.

Reply:
That's a helpful doc. Can we add a link to https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md from the SLI/SLO doc section indicating that the target numbers are assuming you are within the total/cluster/namespace object count bounds?

Reply:
I would like to revise that a bit (and actually we should probably make consistent with cluster/namespace scope terminology). But sure - I agree we should do that.
Let me open an issue for that and I will get back to it once I'm back from vacation.

Reply:
opened https://github.com/kubernetes/community/issues/3659

Reply:
Thanks @wojtek-t, these docs are helpful.

If I understand this right, the SLOs apply to all resource kinds, and the thresholds clarify the supported carnality of each kind, which will be even more clear with the link added. Then for CRDs, how about we also target the same SLOs except with the "+ conversion webhook latency" caveat? I'll write it up and see how it looks.

Reply:
I've rewritten this section in attempt to align it with the native type SLOs and thresholds. PTAL!